source/widgets/widget.c

```diff
@@ -132,308 +154,370 @@ void widget_set_enabled(widget *wid, gboolean enabled) {
+          cairo_line_to(d, wid->w - margin_right - offset_rbr,
+                        wid->h - margin_bottom);
+          cairo_line_to(d, wid->w - margin_right - offset_rbr,
                         wid->h - margin_bottom - bottom);
+          if (radius_inner_br > 0) {
+            cairo_arc_negative(
+                d, wid->w - margin_right - right - radius_inner_br,
+                wid->h - margin_bottom - bottom - radius_inner_br,
+                radius_inner_br, G_PI_2, 0.0);
+            cairo_line_to(d, wid->w - margin_right - right,
+                          wid->h - margin_bottom - offset_bbr);
+          }
+          cairo_line_to(d, wid->w - margin_right,
+                        wid->h - margin_bottom - offset_bbr);
+          cairo_close_path(d);
+          cairo_fill(d);
+        }
+        if (radius_bl > 0) {
+          double radius_outer = radius_bl + minof_bl;
+          cairo_arc(d, margin_left + radius_outer,
+                    wid->h - margin_bottom - radius_outer, radius_outer, G_PI_2,
+                    G_PI);
+          cairo_line_to(d, margin_left, wid->h - margin_bottom - offset_bbl);
+          cairo_line_to(d, margin_left + left,
+                        wid->h - margin_bottom - offset_bbl);
+          if (radius_inner_bl > 0) {
+            cairo_arc_negative(d, margin_left + left + radius_inner_bl,
+                               wid->h - margin_bottom - bottom -
+                                   radius_inner_bl,
+                               radius_inner_bl, G_PI, G_PI_2);
+            cairo_line_to(d, margin_left + offset_lbl,
+                          wid->h - margin_bottom - bottom);
+          }
+          cairo_line_to(d, margin_left + offset_lbl, wid->h - margin_bottom);
+          cairo_close_path(d);
+
+          cairo_fill(d);
         }
-        cairo_line_to(d, margin_left + offset_lbl, wid->h - margin_bottom);
-        cairo_close_path(d);
-
-        cairo_fill(d);
       }
-      cairo_restore(d);
+      #if 0
+       char buff[64];
+       snprintf(buff, 64, "%u", wid->repaint_debug_index);
+       cairo_move_to(d, wid->w / 2, wid->h / 2);
+       cairo_set_source_rgba(d, 1, 1, 1, 1);
+       cairo_show_text(d, buff);
+      #endif
+      cairo_destroy(d);
+      wid->repaint_debug_index++;
     }
+
+    cairo_save(c);
+    cairo_translate(c, wid->x, wid->y);
+    cairo_set_source_surface(c, wid->surf, 0, 0);
+    cairo_paint(c);
+    cairo_restore(c);
   }
 }
 
-void widget_free(widget *wid) {
+/**
+ * reference goes downwards.
+ * Destroying goes upwards.
+ *
+ * This is done because the child often references it parent in the drawing
+ * screen. So the parent cannot dissapear before the child.
+ */
+void widget_ref(widget *wid) {
   if (wid == NULL) {
     return;
   }
+  g_atomic_rc_box_acquire(wid);
+  if ( wid->parent ) {
+    g_atomic_rc_box_acquire(wid->parent);
+  }
+}
+
+void widget_unref(widget *wid) {
+  widget_free(wid);
+  if ( wid->parent ) {
```

Comment:
```suggestion
  if (wid != NULL && wid->parent) {
```

It's possible that it's called with NULL in _rofi_view_reload_widget

Reply:
hmm this is indeed wrong! use after free.

Reply:
Ah yes well... Actually I was focused on the NULL case so my fix is also wrong